internal/zinc-core/src/main/scala/sbt/internal/inc/Stamp.scala

```diff
@@ -58,10 +59,9 @@ trait WithPattern { protected def Pattern: Regex }
 import java.lang.{ Long => BoxedLong }
 
 /** Define the hash of the file contents. It's a typical stamp for compilation sources. */
-final class Hash(val value: Array[Byte]) extends StampBase {
-  val hexHash = IOHash.toHex(value)
+final class Hash(val hexHash: String) extends StampBase {
```

Comment:
@dwijnand You're right, this assumes that the internals will instantiate `Hash` with a hexadecimal value. But this is kind of the point, as this API does not leak to the users and there is only one legit way of creating a hash stamp:
```
  val forHash = (toStamp: File) => tryStamp(new Hash(IOHash.toHex(IOHash(toStamp))))
```
  
... which uses `IOHash.toHex` and makes sure we pass in a valid string to the constructor.
  
The reason why I made this change is because I think comparing bytes is more expensive than strings. `String.==` only checks referential equality, while `sameElements` in an array of bytes causes the comparison of every chunk of data (where chunk is a byte, though it could be a little bit optimized by the JVM): `case (h1: Hash, h2: Hash) => h1.value sameElements h2.value`.
  
What we can do to fix the "type-safety" issue is to make the constructor private and ensure that the instantiation that I show before is the only one that instantiates the class. WDYT?